content/unroll-default-arguments.md

```diff
@@ -0,0 +1,921 @@
+   def this(s: String, n: Int, b: Boolean) = this(s, n, b, 0L)
+   
+   def copy(s: String, n: Int) = copy(s, n, this.b, this.l)
+   def copy(s: String, n: Int, b: Boolean) = copy(s, n, b, this.l)
+   
+   def foo = s + n + b
+}
+object Unrolled{
+   def apply(s: String, n: Int) = apply(s, n, true, 0L)
+   def apply(s: String, n: Int, b: Boolean) = apply(s, n, b, 0L)
+}
+```
+
+Notes:
+
+1. `@unroll`ed `case class`es are fully binary and backwards compatible in Scala 3, but not in Scala 2 
+
+2. `.unapply` does not need to be duplicated in Scala 3.x, as its signature
+   `def unapply(x: Unrolled): Unrolled` does not change when new `case class` fields are
+   added.
+
+3. Even in Scala 2.x, where `def unapply(x: Unrolled): Option[TupleN]` is not
+   binary compatible, pattern matching on `case class`es is already binary compatible 
+   to addition of new fields due to 
+   [Option-less Pattern Matching](https://docs.scala-lang.org/scala3/reference/changed-features/pattern-matching.html).
+   Thus, only calls to `.tupled` or `.curried` on the `case class` companion `object`, or direct calls 
+   to `.unapply` on an unrolled `case class` in Scala 2.x (shown below)
+   will cause a crash if additional fields were added: 
+
+```scala
+def foo(t: (String, Int)) = println(t)
+Unrolled.unapply(unrolled).map(foo)
+```
+
+In Scala 3, `@unroll`ing a `case class` also needs to generate a `fromProduct` 
+implementation in the companion object, as shown below:
+
+```scala
+def fromProduct(p: Product): CaseClass = p.productArity match
+  case 2 =>
+    CaseClass(
+      p.productElement(0).asInstanceOf[...],
+      p.productElement(1).asInstanceOf[...],
+    )
+  case 3 =>
+    CaseClass(
+      p.productElement(0).asInstanceOf[...],
+      p.productElement(1).asInstanceOf[...],
+      p.productElement(2).asInstanceOf[...],
+    )
+  ...
+```
+
+This is not necessary for preserving binary compatibility - the method signature of
+`def fromProduct` does not change depending on the number of fields - but it is
+necessary to preserve semantic compatibility. `fromProduct` by default does not
+take into account field default values, and this change is necessary to make it
+use them when the given `p: Product` has a smaller `productArity` than the current
+`CaseClass` implementation
+
+### Abstract Methods
+
+Apart from `final` methods, `@unroll` also supports purely abstract methods. Consider 
+the following example with a trait `Unrolled` and an implementation `UnrolledObj`:
+
+```scala
+trait Unrolled{ // version 3
+  def foo(s: String, n: Int = 1, @unroll b: Boolean = true, @unroll l: Long = 0): String
+}
+```
+```scala
+object UnrolledObj extends Unrolled{ // version 3
+  def foo(s: String, n: Int = 1, @unroll b: Boolean = true, @unroll l: Long = 0) = s + n + b
+}
+```
+
+This unrolls to:
+```scala
+trait Unrolled{ // version 3
+  def foo(s: String, n: Int = 1, @unroll b: Boolean = true, @unroll l: Long = 0): String = foo(s, n, b)
```

Comment:
This unroll will not cause a `LinkageError`, but it is semantically incorrect anyway. *Ignoring* the value of the parameter at definition site is most likely not allowed by the new spec of the method. It is certainly not at all what a default value stands for at the source level.

Reply:
> but it is semantically incorrect anyway. Ignoring the value of the parameter at definition site is most likely not allowed by the new spec of the method.

I think the first line here doesn't really follow from the second. `@unroll` doesn't make any guarantees over the spec of the method; that's up to you to specify and implement. 

* Some methods' specs will require that you specifically handle the new default parameter. These are fundamental incompatibilities, and `@unroll` can do nothing to help

* Some methods' specs will be fine with you ignoring additional default parameters. For this category, there is no fundamental incompatibility, and `@unroll` makes it so there is no technical/implementation incompatibility either

Now, exactly what the split is between these two categories is an empirical question, but it seems plausible that the latter category is non-zero, and for those this style of reverse forwarder would be useful. But this is up for discussion. If people think this treatment of unrolled abstract methods is unintuitive/confusing/not-worth-the-complexity, I'd be happy to drop it

Reply:
I suggest dropping it. I don't doubt that there are situations where the specs would be fine, but I also expect them to be pretty rare. Regardless, it is infinitely more dangerous than `@unroll` everywhere else, since everywhere else any spec will be accommodated by `@unroll`. I suggest we stick to situations where we can do no harm.

Reply:
One way of thinking about `@unroll`ing abstract methods is that it's meant for adding _optional_ parameters to an invocation _protocol_, rather than any specific caller or implementor.

These are parameters that is optional both for the caller to _pass_, as well as the implementor to _handle_. Beyond this specific case of scala `trait`s with abstract `def`s, this is a relatively common concept in software engineering. 

Examples include:

- HTTP headers, which servers and clients exchange, are usually optional: e.g. `Content-Type` `SHOULD` be provided, not `MUST` [link](https://stackoverflow.com/questions/15860742/is-content-type-mandatory-in-http-post-request), and it is expected that people can extend the HTTP protocol with custom headers that are optional to both clients and servers

- Common invocation protocols like GRPC treat addition of a fields with defaults as backwards compatible ([link](https://learn.microsoft.com/en-us/aspnet/core/grpc/versioning?view=aspnetcore-8.0)), allowing either caller or implementor to omit the field. 
    - The linked documentation for a the .NET GRPC server states "*Adding a field to a request message - Fields added to a request message are deserialized with the default value on the server when not set. To be a non-breaking change, the service must succeed when the new field isn't set by older clients.*". This is similar to the limits around `@unroll`ing abstract method `def`s we have discussed here

- `case class`es themselves are often invocation protocols! 
    - A method taking a single `case class` instance is isomorphic to a method taking a loose parameter list
    - The `case class` definition itself is isomorphic to an abstract method `def`: both specify precisely what typed values the implementor method will have to work with, but now _how_ it will work with those values
    - Thus, adding fields to the `case class` is isomorphic to adding additional parameters to the abstract method `def`
    - If we accept that we should be able to pass newer `case class`es instances with more fields to older versions of code that would simply ignore the newer/unknown fields, why should we reject using a newer abstract method to pass the equivalent loose parameter list to an older version of the concrete implementation, that would ignore the newer/unknown parameters in the same way?

In any protocol-evolution there is a risk of breaking semantic compatibility: it doesn't matter whether you are evolving a Scala abstract-`trait`/implementing-`object`/callsites, or you are evolving a bunch of Protobuf/GRPC schema files and some C++ binary servers and clients that implements it. But the fundamental act of "adding an optional parameter to a protocol that can be ignored by both client or server" is not itself breaking, and in fact seems to be a pretty common pattern/convention in software.

`@unroll` on abstract method `def`s thus matches up with a very common pattern in protocol evolution, just like `@unroll`ing `case class`es matches up with common patterns in schema evolution, and `@unroll`ing concrete method `def`s matches up with common patterns in API evolution. Each of the three challenges can be solved independently of the others, but I think it's a reasonable case to say that solving all three with the same syntax and (almost) the same semantics is a reasonable choice to make